src/main/java/org/jenkinsci/plugins/workflow/steps/scm/SCMStep.java

```diff
@@ -97,14 +98,11 @@ public boolean isChangelog() {
     }
 
     public final void checkout(Run<?,?> run, FilePath workspace, TaskListener listener, Launcher launcher) throws Exception {
-            File changelogFile = null;
+        File tempChangelogFile = null;
+        try {
             if (changelog) {
-                for (int i = 0; ; i++) {
-                    changelogFile = new File(run.getRootDir(), "changelog" + i + ".xml");
-                    if (!changelogFile.exists()) {
-                        break;
-                    }
-                }
+                // We use a temp file so that any failures during the checkout do not leave invalid files in the build directory.
+                tempChangelogFile = Files.createTempFile("changelog", ".xml").toFile();
```

Comment:
This could be on a different filesystem than `run.rootDir`, but you are not passing `ATOMIC_MOVE` to `move`, so I _think_ that will not break. Still, I would suggest making the temp file in the build directory—even if it is invalid, its mere presence is harmless so long as it was not passed to an `SCMListener.onCheckout`. (For that matter, you do not even really have to rename it. There is no particular reason the files need to use a specific naming pattern.)

Reply:
> Still, I would suggest making the temp file in the build directory—even if it is invalid, its mere presence is harmless so long as it was not passed to an SCMListener.onCheckout

In an [earlier comment](https://github.com/jenkinsci/workflow-scm-step-plugin/pull/32#discussion_r249973975) it seemed like you thought that might cause problems:

> The risk is that something goes wrong during checkout and you leave behind an empty (and thus malformed) XML file in the build directory, leading to similar confusing errors

Was your concern just that we need to be careful not to give something a reference an invalid file if the checkout fails, or that any invalid files in the build directory may be a problem?

Reply:
Sorry for the back-and-forth, I have not paid close attention during reviews.

> Was your concern just that we need to be careful not to give something a reference an invalid file if the checkout fails

Yes.

> or that any invalid files in the build directory may be a problem

It seems that they would not be. I am not sure whether the simple system in #31 was safe, but at any rate I think a malformed XML file in the build directory is harmless unless it actually makes its way to an `onCheckout` call, which should not happen if `checkout` threw an exception.

Reply:
Done in 53a57a4, although I decided to delete the file if an exception is caught just in case something else would end up looking at the file.